File: leetcode/facebook/3sum/SolutionTimeout.py
from collections import defaultdict
from sys import maxsize as maxint


class Solution:
    def threeSum(self, nums):
        n = len(nums)
        sum2 = defaultdict(lambda: [])
        minn, maxn = maxint, -maxint
        for i in range(n):
            y = nums[i]
            minn = min(minn, y)
            maxn = max(maxn, y)
            for j in range(i + 1, n):
                sum2[y + nums[j]].append((i, j))
        base = maxn - minn + 1

        def calc_key(x, y, z):
            arr = sorted((x, y, z))
            key = 0
            for i in range(2):
                key += (base ** i) * arr[i]
            return key

        triples = []
        used = dict()
        for k in range(n):
            x = nums[k]
            if -x in sum2:
                for i, j in sum2[-x]:
                    if k == i or k == j:
                        continue
                    y, z = nums[i], nums[j]
                    key = calc_key(x, y, z)
                    if key not in used:
                        used[key] = True
                        triples.append((x, y, z))
        return triples

File: leetcode/second-pass/facebook/3sum/SolutionTimeout.java
class Solution {
    private int[] arr = new int[3];
    private int base;

    private int calcKey(int base, int x, int y, int z) {
        arr[0] = x;
        arr[1] = y;
        arr[2] = z;
        Arrays.sort(arr);
        int key = 0;
        int powBase = 1;
        for(int i=0; i<2; i++) {
            key += powBase * arr[i];
            powBase *= base;
        }
        return key;
    }

    private Map<Integer, List<int[]>> getSum2SetBase(int[] nums) {
        int n = nums.length;
        Map<Integer, List<int[]>> sum2 = new HashMap<>();
        int minn = Integer.MAX_VALUE;
        int maxn = Integer.MIN_VALUE;
        for(int i=0; i<n; i++) {
            int y = nums[i];
            minn = Math.min(minn, y);
            maxn = Math.max(maxn, y);
            for(int j=i+1; j<n; j++) {
                int s = y + nums[j];
                List<int[]> pairs = sum2.getOrDefault(s, new ArrayList<>());
                pairs.add(new int[]{i, j});
                sum2.put(s, pairs);
            }
        }
        base = maxn - minn + 1;
        return sum2;
    }

    public List<List<Integer>> threeSum(int[] nums) {
        Map<Integer, List<int[]>> sum2 = getSum2SetBase(nums);
        List<List<Integer>> triples = new ArrayList<>();
        Set<Integer> used = new HashSet<>();
        for(int k=0; k<nums.length; k++) {
            int x = nums[k];
            if (sum2.containsKey(-x)) {
                for(int[] pair: sum2.get(-x)) {
                    int i = pair[0];
                    int j = pair[1];
                    if (k == i || k == j)
                        continue;
                    int y = nums[i];
                    int z = nums[j];
                    int key = calcKey(base, x, y, z);
                    if (!used.contains(key)) {
                        used.add(key);
                        triples.add(Arrays.asList(x, y, z));
                    }
                }
            }
        }
        return triples;
    }
}